Adding density material to PorousFlow, closes #9027

hsheldon opened the pull request:
Added density material to PorousFlow that enables density varying with porosity
#9027

hsheldon committed:
Adding density material to PorousFlow, closes #9027

hsheldon committed:
Adding density material to PorousFlow, closes #9027

moosebuild commented:
Job `Precheck:linux-gnu` on dcb1dcf wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format c950de054411525efaa3cc3c2781a2734a7c9962`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9080/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9080/style.patch | git apply -v`

permcody commented:
@WilkAndy - This looks fine to me. I'll let you have a look over it though before it gets merged.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h:
```diff
@@ -0,0 +1,72 @@
+   * @param ph phase index
+   * @return fluid density
+   */
+  virtual Real rho_f(unsigned int ph = 0) const;
```
I think change to

```
virtual Real rho_f() const;
```
only.  I can't foresee ever needing the phase number as an input.  If i'm wrong then it's easily changed, but i think making this general is unnecessary presently.  You may disagree!

hsheldon replied:
ok

WilkAndy replied:
Oh, i just realised this function and drho_f_dvar should really go into the base class (TotalGravitationalDensityBase) as virtual pure functions if you want to override them in multiple classes.  Ie, in the TotalGravitationDensityBase.h you'd have
```
virtual Real rho_f() const = 0;
virtual Real drho_f_dvar(unsigned int v) const = 0;
```
and then you'd override as usual in your "FullySaturated" version.  **However** now i'm wondering whether the "override-these-functions" really is a wise design.  I think we're being unnecessarily complicated, and i think just remove these rho_f and drho_f_dvar functions altogether, and just have 

```
_gravdensity[_qp] = _rho_s * (1.0 - _porosity_qp[_qp]) + _rho_f_qp[_qp][0] * _porosity_qp[_qp];
```
in your computeQpProperties.  It's simpler, which is usually better.  If someone really wants to do some fancy overriding in the future then let them and don't imagine you know what they'll really want.  I'm sorry, i know i suggested these functions in the beginning, but i think i was wrong.

WilkAndy commented on modules/porous_flow/include/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.h:
```diff
@@ -0,0 +1,72 @@
+   * @param v variable index
+   * @return derivatives of fluid density
+   */
+  virtual Real drho_f_dvar(unsigned int ph = 0, unsigned int v = 0) const;
```
Remove the
```
unsigned int ph = 0
```
argument, and remove the default from v, ie just make it
```
unsigned int v
```
There's no need for the default.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C:
```diff
@@ -0,0 +1,75 @@
+{
+  InputParameters params = validParams<PorousFlowTotalGravitationalDensityBase>();
+  params.addRequiredRangeCheckedParam<Real>(
+      "rho_s", "rho_s > 0", "The density of the solid matrix");
```
I'm not sure if you might want
```
rho_s >= 0
```
here instead.  I know that in reality rho_s must be positive, but i've been caught in the past when people have suddenly wanted to do some comparison with analytical solutions where some quantity has been ignored so they've wanted to set it zero, and my range check has prevented them.  Probably i'm being overly cautious here - i'll let you decide.

hsheldon replied:
Good point, I did wonder about that.

hsheldon replied:
Done

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C:
```diff
@@ -0,0 +1,75 @@
+Real
+PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity::rho_f(unsigned int ph) const
+{
+  return _rho_f_qp[_qp][ph];
```
If you follow my suggestion above, this line will just be
```
return _rho_f_qp[_qp][0];
```
in your case, and it'll be ready to be overridden by someone else.

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C:
```diff
@@ -0,0 +1,75 @@
+  if (_nodal_material == true)
+    mooseError("PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity is only defined for "
+               "at_nodes = false");
+}
```
Probably you should check
```
_num_phases == 1
```
to prevent people from doing something too silly.

hsheldon replied:
Done

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C:
```diff
@@ -0,0 +1,75 @@
+PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity::drho_f_dvar(unsigned int ph,
+                                                                           unsigned int v) const
+{
+  return _drho_f_qp_dvar[_qp][ph][v];
```
and this line will be
```
return _drho_f_qp_dvar[_qp][0][v];
```

WilkAndy commented on modules/porous_flow/tests/density/GravDensity01.i:
```diff
@@ -0,0 +1,149 @@
+[]
+
+[GlobalParams]
+  block = 0
```
block=0 is probably unnecessary

WilkAndy commented on modules/porous_flow/tests/density/GravDensity01.i:
```diff
@@ -0,0 +1,149 @@
+  [../]
+[]
+
+[Functions]
```
No [Functions]

WilkAndy commented on modules/porous_flow/tests/density/GravDensity01.i:
```diff
@@ -0,0 +1,149 @@
+[Executioner]
+  solve_type = Newton
+  type = Steady
+  l_tol = 1E-5
```
almost all these Executioner parameters are unneeded

WilkAndy commented on modules/porous_flow/tests/density/GravDensity01.i:
```diff
@@ -0,0 +1,149 @@
+[]
+
+[AuxVariables]
+  [./density]
```
it would be good to check the derivatives too.  i know that no kernel is actually using them, but it'd still be good.  Use a MaterialStdVectorAux

WilkAndy replied:
In which case, having a lower bulk density will be a good idea.  And maybe a more complicated porosity function - i'll let you decide as this could get really complicated quite quickly!

hsheldon replied:
Hmm yes. The derivatives should be zero as it stands at the moment, because porosity is constant. So yes I would have to put in a more complicated porosity function to make the  derivative check meaningful.

hsheldon replied:
As discussed, you are correct there will be a non-zero derivative wrt fluid density. I will implement a check for that.

WilkAndy commented on modules/porous_flow/tests/density/tests:
```diff
@@ -0,0 +1,8 @@
+    type = 'CSVDiff'
+    input = 'GravDensity01.i'
+    csvdiff = 'GravDensity01.csv'
+    rel_err = 1.0E-8
```
probably don't need the rel_err parameter

WilkAndy reviewed (changes requested):
A few minor changes

hsheldon committed:
Making some changes suggested by Andy:
- Removing rho_f functions
- Adding a check of derivatives to the test and reducing magnitude of density terms
- Removing rel_err

moosebuild commented:
Job `Precheck:linux-gnu` on 1e641a9 wanted to post the following:

Your code requires style changes.
In the top level of your repository, you can run:

`git clang-format c950de054411525efaa3cc3c2781a2734a7c9962`

Alternatively, apply the patch found [here](http://mooseframework.org/docs/PRs/9080/style.patch).
You can directly apply the patch by running:

`curl -s http://mooseframework.org/docs/PRs/9080/style.patch | git apply -v`

WilkAndy commented on modules/porous_flow/tests/density/GravDensity01.i:
```diff
@@ -1,8 +1,14 @@
+#   = phi * rho_f / B
+#   where rho_f = rho_0 * exp(pp / B) is fluid density, pp is fluid pressure, phi is
+#   porosity and B is fluid bulk modulus
+# With pp = 0, d_rho / d_pp = phi * rho_0 / B = 0.1 * 2 / 4 = 0.05
```
Probably you have to "git add" the changed csv file too

hsheldon replied:
Yes.... I shouldn't have pushed yet, still work in progress.

hsheldon replied:
All good now, I hope!

hsheldon committed:
merging after rebase

WilkAndy commented on modules/porous_flow/src/materials/PorousFlowTotalGravitationalDensityFullySaturatedFromPorosity.C:
```diff
@@ -0,0 +1,66 @@
+    mooseError("The Dictator proclaims that the number of phases is ",
+               _dictator.numPhases(),
+               " whereas PorousFlowThermalConductivityFromPorosity can only be used for 1-phase "
+               "simulations. Be aware that the Dictator has noted your mistake.");
```
Excellent mooseError message!

hsheldon replied:
I think you wrote that originally, I copied it from elsewhere in PorousFlow!

WilkAndy reviewed (approved):
Looks good now.  Thanks @hsheldon.  As far as i'm concerned this can be merged if it passes the tests.

moosebuild commented:
Job `Moose Documentation:linux-clang` on 3e64d50 wanted to post the following:

View the site at http://mooseframework.org/docs/PRs/9080/site
